### PR TITLE
refactor: Remove uses of `std::enable_if`

### PR DESCRIPTION
**Pull request by stephenswat**
With the move to C++20, we no longer need `std::enable_if` as we can use the much more elegant `requires` keyword. This commit swaps virtually all uses of `std::enable_if`, except those in the existing pre-C++20 concepts library.

### CODE REVIEW

**Comment by stephenswat** [comment]
This was quite meditative.

File: Core/include/Acts/EventData/TrackStateProxy.hpp [hunk 1]
```diff
@@ -386,8 +386,8 @@ class TrackStateProxy {
   /// This overloaded is only enabled if not read-only, and returns a mutable
   /// reference.
   /// @return Mutable reference to the pathlength.
-  template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
-  double& pathLength() {
+  template <bool RO = ReadOnly>
+  requires(!RO) double& pathLength() {
```
Comment (paulgessinger) [inline Core/include/Acts/EventData/TrackStateProxy.hpp:390]: Seems like I missed these though 😅 
```suggestion
  requires(!ReadOnly) double& pathLength() {
```

Comment (paulgessinger) [reply Core/include/Acts/EventData/TrackStateProxy.hpp:390]: Aside from that, in the rest of the changes I've made so far, I've put the `requires` clause behind the function name. Maybe we can keep that consistent?

**Review by paulgessinger — COMMENTED** [review]
I like it.

1. Can you remove the cases where function template arguments are no longer necessary?
2. Can you move the `requires` clause behind the function name and argument list?

**Comment by stephenswat** [comment]
@paulgessinger this is now all requires-east (requires after the function name) and I removed all the unnecessary type aliases.

File: Core/include/Acts/Utilities/VectorHelpers.hpp [hunk 2]
```diff
@@ -55,9 +47,12 @@ double phi(const Eigen::MatrixBase<Derived>& v) noexcept {
 /// @tparam T anything that has a phi method
 /// @param v Any type that implements a phi method
 /// @return The phi value
-template <typename T,
-          std::enable_if_t<detail::has_phi_method<T>::value, int> = 0>
-double phi(const T& v) noexcept {
+template <typename T>
+double phi(const T& v) noexcept
+  requires requires(const T& w) {
+    { w.phi() };
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/VectorHelpers.hpp:53]: Yes! Although, annoyingly, clang-tidy will complain if variables in your required clause clash name-wise with the function arguments. But I guess that's fine.

I wonder if you can reuse variable names in your requires clause?

File: Core/include/Acts/Utilities/Axis.hpp [hunk 3]
```diff
@@ -174,11 +174,12 @@ class Axis<AxisType::Equidistant, bdt> final : public IAxis {
...
   ///       as neighbors
-  template <AxisBoundaryType T = bdt,
-            std::enable_if_t<T == AxisBoundaryType::Open, int> = 0>
+  template <AxisBoundaryType T = bdt>
   NeighborHoodIndices neighborHoodIndices(std::size_t idx,
-                                          std::pair<int, int> sizes = {
-                                              -1, 1}) const {
+                                          std::pair<int, int> sizes = {-1,
+                                                                       1}) const
+    requires(T == AxisBoundaryType::Open)
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/Axis.hpp]: Fixed.

File: Core/include/Acts/Utilities/Axis.hpp [hunk 4]
```diff
@@ -196,11 +197,12 @@ class Axis<AxisType::Equidistant, bdt> final : public IAxis {
   /// @return Set of neighboring bin indices (global)
   /// @note Bound varies given bin and allows 1 and NBins (regular bins)
   ///       as neighbors
-  template <AxisBoundaryType T = bdt,
-            std::enable_if_t<T == AxisBoundaryType::Bound, int> = 0>
+  template <AxisBoundaryType T = bdt>
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/Axis.hpp]: Fixed.

File: Core/include/Acts/Utilities/Delegate.hpp [hunk 5]
```diff
@@ -65,11 +64,12 @@ class Delegate<R(Args...), H, O> {
...
   using NonOwningDelegate =
       Delegate<R(Args...), holder_type, DelegateType::NonOwning>;
+
   template <typename T>
-  using isNoFunPtr =
-      std::enable_if_t<!std::is_convertible_v<std::decay_t<T>, function_type> &&
-                       !std::is_same_v<std::decay_t<T>, OwningDelegate> &&
-                       !std::is_same_v<std::decay_t<T>, NonOwningDelegate>>;
+  using isNoFunPtr = std::conjunction<
+      std::negation<std::is_convertible<std::decay_t<T>, function_type>>,
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/Delegate.hpp:70]: Yes, because `isNoFunPtr` is a type, not a boolean value. So you can't actually negate it using `!`, you need to use `std::negation`. I wanted to make this a proper concept at some point, but I was having some trouble with it (don't remember what exactly) so I gave up on that for now.

File: Core/include/Acts/Utilities/VectorHelpers.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/VectorHelpers.hpp:53]: Holy shit https://godbolt.org/z/E4Mv4a9jv

File: Core/include/Acts/Utilities/Delegate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/Delegate.hpp:70]: Actually, let me try again if I can conceptify this properly.

File: Core/include/Acts/Utilities/VectorHelpers.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/Utilities/VectorHelpers.hpp:53]: nice! `requires requires` still confuses me 😄

File: Core/include/Acts/Utilities/Delegate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/Delegate.hpp:70]: Ah! I remember; this was because `isNoFunPtr` binds a lot of typedefs for the class it lives in, which you are allowed to do for old-school trait classes but you cannot define concepts inside of types, so the concept would need to have way more arguments.

File: Core/include/Acts/Utilities/VectorHelpers.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/VectorHelpers.hpp:53]: I think at this point we're all still learning how to work with concepts effectively :smile:

[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/VectorHelpers.hpp]: Updated the PR to use this somewhat less-verbose syntax.

[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/Utilities/VectorHelpers.hpp]: Also, this is a good reason to have the requires clause after the function arguments!

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [9589527d0766e0a52aa97a26dbaabb622dce3378]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42718012) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7994592)!
